File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

/*
 use hash to storage all brackets, open brackets as key, closing brackets as value
literate s, if current char is open brackets store it in stack arr
if current char is closing brackets, pop the last ele in stack and see are they matching bracket, if so, continue if not, return false
return stack.length is empty or not
 */
var isValid = function(s) {
   
    const brackets = {
        "{" : "}", 
        "[" : "]",
        "(" : ")"
    }
    const stack = [s[0]]
    for (let i = 1; i < s.length; i++) {
        const curr = s[i]
        if (brackets.hasOwnProperty(curr)) {
            stack.push(curr)
        } else {
            const last = stack.pop()
            if (brackets[last] !== curr) return false
        }
    }
    return stack.length === 0
};
    
    

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
    
        dict = {
                "}" : "{" ,
                "]" : "[" ,
                ")" : "("
               }
        
        stack = []
        
        for char in s:
            if char in dict:
                top_ele = stack.pop() if stack else '#'
            
                if dict[char] != top_ele:
                    return False
            else:
                stack.append(char)
            
        return not stack